File: Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        std::list<TreeNode*> workingList;
        // false: from left to right
        // true: from right to left
        bool direction = false;
        std::vector<std::vector<int> > res;
        if(!root){
            // Empty input expects one-dimensional empty output
            //res.push_back(std::vector<int>());
            return res;
        }
        
        workingList.push_back(root);
        while(!workingList.empty()){
            std::list<TreeNode*> newList;
            std::vector<int> newElement;
            if(!direction){
                for(std::list<TreeNode*>::iterator it = workingList.begin(); it != workingList.end(); ++it){
                    TreeNode* curr = (*it);
                    newElement.push_back(curr->val);
                    if(curr->left)      newList.push_back(curr->left);
                    if(curr->right)     newList.push_back(curr->right);
                }
            }
            else{
                for(std::list<TreeNode*>::reverse_iterator it = workingList.rbegin(); it != workingList.rend(); ++it){
                    TreeNode* curr = (*it);
                    newElement.push_back(curr->val);
                    if(curr->right)     newList.push_front(curr->right);
                    if(curr->left)      newList.push_front(curr->left);
                }
            }
            direction = !direction;
            workingList = newList;
            newList.clear();
            res.push_back(newElement);
        }
        
        return res;
    }
};

File: Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        
        if(root == null){
            return ret;
        }
        
        // false: currently from left to right
        // true: currently from right to left
        boolean direction = false;
        Stack<TreeNode> active = new Stack<TreeNode>();
        active.push(root);
        
        while(!active.empty()){
            Stack<TreeNode> next = new Stack<TreeNode>();
            List<Integer> element = new ArrayList<Integer>();
            while(!active.empty()){
                TreeNode temp = active.pop();
                element.add(temp.val);
                if(direction){
                    if(temp.right != null){
                        next.add(temp.right);
                    }
                    if(temp.left != null){
                        next.add(temp.left);
                    }
                }
                else{
                    if(temp.left != null){
                        next.add(temp.left);
                    }
                    if(temp.right != null){
                        next.add(temp.right);
                    }
                }
            }
            direction = !direction;
            active = next;
            ret.add(element);
        }
        
        return ret;
    }
}
